File: frontend/src/api/deposit-slots.ts
import { apiClient } from './client';
import type {
  DepositSlot,
  DepositSlotListResponse,
  CreateDepositSlotRequest,
  UpdateDepositSlotRequest,
} from '@/types';

/**
 * API response type after Axios interceptor transforms keys to camelCase.
 */
interface DepositSlotApiResponse {
  id: string;
  editionId: string;
  startDatetime: string;
  endDatetime: string;
  maxCapacity: number;
  reservedForLocals: boolean;
  description: string | null;
  registeredCount: number;
  createdAt: string;
}

interface DepositSlotListApiResponse {
  items: DepositSlotApiResponse[];
  total: number;
}

/**
 * Transform API response to frontend DepositSlot type.
 * Axios interceptor already converts snake_case to camelCase.
 */
function transformDepositSlot(data: DepositSlotApiResponse): DepositSlot {
  return {
    id: data.id,
    editionId: data.editionId,
    startDatetime: data.startDatetime,
    endDatetime: data.endDatetime,
    maxCapacity: data.maxCapacity,
    reservedForLocals: data.reservedForLocals,
    description: data.description,
    registeredCount: data.registeredCount ?? 0,
    createdAt: data.createdAt,
  };
}

/**
 * Deposit slots API endpoints.
 */
export const depositSlotsApi = {
  /**
   * List deposit slots for an edition.
   */
  getDepositSlots: async (editionId: string): Promise<DepositSlotListResponse> => {
    const response = await apiClient.get<DepositSlotListApiResponse>(
      `/v1/editions/${editionId}/deposit-slots`
    );
    return {
      items: response.data.items.map(transformDepositSlot),
      total: response.data.total,
    };
  },

  /**
   * Get a single deposit slot by ID.
   */
  getDepositSlot: async (editionId: string, slotId: string): Promise<DepositSlot> => {
    const response = await apiClient.get<DepositSlotApiResponse>(
      `/v1/editions/${editionId}/deposit-slots/${slotId}`
    );
    return transformDepositSlot(response.data);
  },

  /**
   * Create a new deposit slot.
   */
  createDepositSlot: async (
    editionId: string,
    data: CreateDepositSlotRequest
  ): Promise<DepositSlot> => {
    const response = await apiClient.post<DepositSlotApiResponse>(
      `/v1/editions/${editionId}/deposit-slots`,
      {
        start_datetime: data.startDatetime,
        end_datetime: data.endDatetime,
        max_capacity: data.maxCapacity ?? 20,
        reserved_for_locals: data.reservedForLocals ?? false,
        description: data.description,
      }
    );
    return transformDepositSlot(response.data);
  },

  /**
   * Update a deposit slot.
   */
  updateDepositSlot: async (
    editionId: string,
    slotId: string,
    data: UpdateDepositSlotRequest
  ): Promise<DepositSlot> => {
    const payload: Record<string, unknown> = {};

    if (data.startDatetime !== undefined) payload.start_datetime = data.startDatetime;
    if (data.endDatetime !== undefined) payload.end_datetime = data.endDatetime;
    if (data.maxCapacity !== undefined) payload.max_capacity = data.maxCapacity;
    if (data.reservedForLocals !== undefined) payload.reserved_for_locals = data.reservedForLocals;
    if (data.description !== undefined) payload.description = data.description;

    const response = await apiClient.put<DepositSlotApiResponse>(
      `/v1/editions/${editionId}/deposit-slots/${slotId}`,
      payload
    );
    return transformDepositSlot(response.data);
  },

  /**
   * Delete a deposit slot.
   */
  deleteDepositSlot: async (editionId: string, slotId: string): Promise<void> => {
    await apiClient.delete(`/v1/editions/${editionId}/deposit-slots/${slotId}`);
  },
};

File: backend/app/api/v1/endpoints/deposit_slots.py
"""Deposit slot API endpoints."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from app.dependencies import DBSession, require_role
from app.exceptions import EditionNotFoundError, ValidationError
from app.models import User
from app.repositories import DepositSlotRepository
from app.repositories.edition_depositor import EditionDepositorRepository
from app.schemas import (
    DepositSlotCreate,
    DepositSlotListResponse,
    DepositSlotResponse,
    DepositSlotUpdate,
)
from app.services import EditionService

router = APIRouter()


def get_edition_service(db: DBSession) -> EditionService:
    """Get EditionService instance."""
    return EditionService(db)


def get_deposit_slot_repository(db: DBSession) -> DepositSlotRepository:
    """Get DepositSlotRepository instance."""
    return DepositSlotRepository(db)


EditionServiceDep = Annotated[EditionService, Depends(get_edition_service)]
DepositSlotRepoDep = Annotated[DepositSlotRepository, Depends(get_deposit_slot_repository)]


@router.get(
    "",
    response_model=DepositSlotListResponse,
    summary="List deposit slots",
    description="List all deposit slots for an edition.",
)
async def list_deposit_slots(
    edition_id: str,
    db: DBSession,
    edition_service: EditionServiceDep,
    repo: DepositSlotRepoDep,
    current_user: Annotated[User, Depends(require_role(["manager", "administrator"]))],
):
    """List all deposit slots for an edition."""
    # Verify edition exists
    try:
        await edition_service.get_edition(edition_id)
    except EditionNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Edition {edition_id} not found",
        )

    slots, total = await repo.list_by_edition(edition_id)

    ed_repo = EditionDepositorRepository(db)
    items = []
    for s in slots:
        data = DepositSlotResponse.model_validate(s)
        data.registered_count = await ed_repo.count_by_slot(s.id)
        items.append(data)

    return DepositSlotListResponse(
        items=items,
        total=total,
    )


@router.post(
    "",
    response_model=DepositSlotResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create deposit slot",
    description="Create a new deposit slot for an edition.",
)
async def create_deposit_slot(
    edition_id: str,
    request: DepositSlotCreate,
    edition_service: EditionServiceDep,
    repo: DepositSlotRepoDep,
    current_user: Annotated[User, Depends(require_role(["manager", "administrator"]))],
):
    """Create a new deposit slot."""
    # Verify edition exists and is editable
    try:
        edition = await edition_service.get_edition(edition_id)
    except EditionNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Edition {edition_id} not found",
        )

    if edition.is_closed:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Cannot modify a closed edition",
        )

    # Check for overlapping slots
    has_overlap = await repo.has_overlapping_slot(
        edition_id=edition_id,
        start_datetime=request.start_datetime,
        end_datetime=request.end_datetime,
    )

    if has_overlap:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Ce créneau chevauche un créneau existant",
        )

    slot = await repo.create(
        edition_id=edition_id,
        start_datetime=request.start_datetime,
        end_datetime=request.end_datetime,
        max_capacity=request.max_capacity,
        reserved_for_locals=request.reserved_for_locals,
        description=request.description,
    )

    return DepositSlotResponse.model_validate(slot)


@router.get(
    "/{slot_id}",
    response_model=DepositSlotResponse,
    summary="Get deposit slot",
    description="Get a single deposit slot by ID.",
)
async def get_deposit_slot(
    edition_id: str,
    slot_id: str,
    edition_service: EditionServiceDep,
    repo: DepositSlotRepoDep,
    current_user: Annotated[User, Depends(require_role(["manager", "administrator"]))],
):
    """Get a single deposit slot."""
    # Verify edition exists
    try:
        await edition_service.get_edition(edition_id)
    except EditionNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Edition {edition_id} not found",
        )

    slot = await repo.get_by_id(slot_id)

    if not slot or slot.edition_id != edition_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Deposit slot {slot_id} not found",
        )

    return DepositSlotResponse.model_validate(slot)


@router.put(
    "/{slot_id}",
    response_model=DepositSlotResponse,
    summary="Update deposit slot",
    description="Update a deposit slot.",
)
async def update_deposit_slot(
    edition_id: str,
    slot_id: str,
    request: DepositSlotUpdate,
    edition_service: EditionServiceDep,
    repo: DepositSlotRepoDep,
    current_user: Annotated[User, Depends(require_role(["manager", "administrator"]))],
):
    """Update a deposit slot."""
    # Verify edition exists and is editable
    try:
        edition = await edition_service.get_edition(edition_id)
    except EditionNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Edition {edition_id} not found",
        )

    if edition.is_closed:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Cannot modify a closed edition",
        )

    slot = await repo.get_by_id(slot_id)

    if not slot or slot.edition_id != edition_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Deposit slot {slot_id} not found",
        )

    # Check for overlapping slots if dates are changing
    if request.start_datetime or request.end_datetime:
        new_start = request.start_datetime or slot.start_datetime
        new_end = request.end_datetime or slot.end_datetime

        has_overlap = await repo.has_overlapping_slot(
            edition_id=edition_id,
            start_datetime=new_start,
            end_datetime=new_end,
            exclude_slot_id=slot_id,
        )

        if has_overlap:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Ce créneau chevauche un créneau existant",
            )

    # Build update kwargs
    update_data = request.model_dump(exclude_unset=True)
    updated_slot = await repo.update(slot, **update_data)

    return DepositSlotResponse.model_validate(updated_slot)


@router.delete(
    "/{slot_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete deposit slot",
    description="Delete a deposit slot.",
)
async def delete_deposit_slot(
    edition_id: str,
    slot_id: str,
    edition_service: EditionServiceDep,
    repo: DepositSlotRepoDep,
    current_user: Annotated[User, Depends(require_role(["manager", "administrator"]))],
):
    """Delete a deposit slot."""
    # Verify edition exists and is editable
    try:
        edition = await edition_service.get_edition(edition_id)
    except EditionNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Edition {edition_id} not found",
        )

    if edition.is_closed:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Cannot modify a closed edition",
        )

    slot = await repo.get_by_id(slot_id)

    if not slot or slot.edition_id != edition_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Deposit slot {slot_id} not found",
        )

    await repo.delete(slot)
